lib/nmatrix/math.rb

```diff
@@ -227,6 +227,27 @@ def factorize_lu with_permutation_matrix=nil
     [t.transpose, FactorizeLUMethods.permutation_matrix_from(pivot)]
   end
 
+  # Reduce self to upper hessenberg form using householder transforms.
+  # 
+  # == References
+  #
+  # * http://en.wikipedia.org/wiki/Hessenberg_matrix
+  # * http://www.mymathlib.com/c_source/matrices/eigen/hessenberg_orthog.c
+  def hessenberg
+    raise ShapeError, "Trying to reduce non 2D matrix to hessenberg form" if 
+      shape.size != 2
+    raise ShapeError, "Trying to reduce non-square matrix to hessenberg form" if 
+      shape[0] != shape[1]
+    raise StorageTypeError, "Matrix must be dense" if stype != :dense
+    raise TypeError, "Works with float matrices only" unless 
+      [:float64,:float32].include?(dtype)
+
+    a = self.clone
+    __hessenberg__(a)
```

Comment:
Might also be nice to have a `hessenberg!` that does this without the clone too.